temporal/tracing.go

```diff
@@ -1,11 +1,136 @@
+		Result   T
+		Err      error
+		Attempts int
+		Status   string
+	}
+)
+
+const tracingKey = "armory-tracing"
+
+func (w workflowTraceParametersPropagator) Inject(ctx context.Context, writer workflow.HeaderWriter) error {
+	kvp, ok := ctx.Value(workflowTraceParametersKey{}).(*map[string]string)
+	if ok && nil != kvp {
+		payload, err := converter.GetDefaultDataConverter().ToPayload(*kvp)
+		if err != nil {
+			return err
+		}
+		writer.Set(tracingKey, payload)
+	}
+	return nil
+}
+
+func (w workflowTraceParametersPropagator) Extract(ctx context.Context, reader workflow.HeaderReader) (context.Context, error) {
+	if raw, ok := reader.Get(tracingKey); ok {
+		var kvp map[string]string
+		if err := converter.GetDefaultDataConverter().FromPayload(raw, &kvp); err != nil {
+			return ctx, nil
+		}
+		ctx = context.WithValue(ctx, workflowTraceParametersKey{}, &kvp)
+	}
+	return ctx, nil
+}
+
+func (w workflowTraceParametersPropagator) InjectFromWorkflow(ctx workflow.Context, writer workflow.HeaderWriter) error {
+	kvp, ok := ctx.Value(workflowTraceParametersKey{}).(*map[string]string)
+	if ok {
+		payload, err := converter.GetDefaultDataConverter().ToPayload(*kvp)
+		if err != nil {
+			return err
+		}
+		writer.Set(tracingKey, payload)
+	}
+	return nil
+}
+
+func (w workflowTraceParametersPropagator) ExtractToWorkflow(ctx workflow.Context, reader workflow.HeaderReader) (workflow.Context, error) {
+	if raw, ok := reader.Get(tracingKey); ok {
+		var kvps map[string]string
+		if err := converter.GetDefaultDataConverter().FromPayload(raw, &kvps); err != nil {
+			return ctx, nil
+		}
+		ctx = workflow.WithValue(ctx, workflowTraceParametersKey{}, &kvps)
+	}
+	return ctx, nil
+}
+
 func newOtelInterceptor() (interceptor.Interceptor, error) {
 	// Uses global tracer provider.
 	return temporalotel.NewTracingInterceptor(temporalotel.TracerOptions{})
 }
+
+func NewWorkflowTraceParametersPropagator() workflow.ContextPropagator {
+	return &workflowTraceParametersPropagator{}
+}
+
+func WithTraceParameters(ctx context.Context, entries ...string) context.Context {
+	container := getOrCreateTags(ctx.Value)
+	return context.WithValue(ctx, workflowTraceParametersKey{}, makeTraceabilityTags(container, entries))
+}
+
+func WithWorkflowTraceParameters(ctx workflow.Context, entries ...string) workflow.Context {
+	container := getOrCreateTags(ctx.Value)
+	return workflow.WithValue(ctx, workflowTraceParametersKey{}, makeTraceabilityTags(container, entries))
+}
+
+func getOrCreateTags(contextGetter func(key any) any) *map[string]string {
+	existing, ok := contextGetter(workflowTraceParametersKey{}).(*map[string]string)
+	return lo.IfF(ok && existing != nil, func() *map[string]string {
+		copy := map[string]string{}
+		for k, v := range *existing {
+			copy[k] = v
+		}
+		return &copy
+	}).ElseF(func() *map[string]string {
+		return &map[string]string{}
+	})
+}
+
+func makeTraceabilityTags(tags *map[string]string, entries []string) *map[string]string {
```

Comment:
This function is called by public functions (`WithTraceParameters`, `WithWorkflowTraceParameters`) but calls a `panic`. I would try to return or log an error instead of panicking here.

https://github.com/golang/go/wiki/CodeReviewComments#dont-panic

Reply:
updated